pkg/liqo-controller-manager/resourceoffer-controller/resourceoffer_controller.go

```diff
@@ -139,14 +149,74 @@ func (r *ResourceOfferReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ResourceOfferReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// select replicated resources only
 	p, err := predicate.LabelSelectorPredicate(crdreplicator.ReplicatedResourcesLabelSelector)
 	if err != nil {
 		klog.Error(err)
 		return err
 	}
 
+	// select virtual kubelet deployments only
+	deployPredicate, err := predicate.LabelSelectorPredicate(metav1.LabelSelector{
+		MatchLabels: vkMachinery.KubeletBaseLabels,
+	})
+	if err != nil {
+		klog.Error(err)
+		return err
+	}
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&sharingv1alpha1.ResourceOffer{}, builder.WithPredicates(p)).
-		Owns(&v1.Deployment{}).
+		Watches(&source.Kind{Type: &v1.Deployment{}},
+			getVirtualKubeletEventHandler(), builder.WithPredicates(deployPredicate)).
 		Complete(r)
 }
+
+// getVirtualKubeletEventHandler creates and returns an event handle with the same behavior of the
+// owner reference event handler, but using an annotation. This allows us to have a graceful deletion
+// of the owned object, impossible using a standard owner reference, keeping the possibility to be
+// triggered on children updates and to enforce their status.
+func getVirtualKubeletEventHandler() handler.EventHandler {
+	return &handler.Funcs{
+		CreateFunc: func(ce event.CreateEvent, rli workqueue.RateLimitingInterface) {
+			if req, err := getRequestFromObject(ce.Object); err == nil {
+				rli.Add(req)
+			}
+		},
+		UpdateFunc: func(ue event.UpdateEvent, rli workqueue.RateLimitingInterface) {
+			if req, err := getRequestFromObject(ue.ObjectOld); err == nil {
+				rli.Add(req)
+			}
+			if req, err := getRequestFromObject(ue.ObjectNew); err == nil {
+				rli.Add(req)
+			}
+		},
+		DeleteFunc: func(de event.DeleteEvent, rli workqueue.RateLimitingInterface) {
+			if req, err := getRequestFromObject(de.Object); err == nil {
+				rli.Add(req)
+			}
+		},
+		GenericFunc: func(ge event.GenericEvent, rli workqueue.RateLimitingInterface) {
+			if req, err := getRequestFromObject(ge.Object); err == nil {
+				rli.Add(req)
+			}
+		},
+	}
+}
+
+// getRequestFromObject reads an annotation in the object and returns the request
+// to be enqueued for the reconciliation.
+func getRequestFromObject(obj client.Object) (reconcile.Request, error) {
```

Comment:
```suggestion
func getResourceRequestFromObject(obj client.Object) (reconcile.Request, error) {
```

I would personally be more explicit, otherwise the name might be misleading.

Reply:
Request in my mind was the `reconcile.Request` 😅

I have definitely to be clearer